### ISSUE TITLE
ROL:  Method of Moving Asymptotes

### ISSUE THREAD

**Issue by Adrian-Diaz** [trilinos/Trilinos#11810]
## Question
Greetings,

I was looking to run a comparison against the MMA method and noticed ROL has an objective interface implementing it. Has this interface been tested successfully before? if so with what algorithms? Just wanted to make sure before I adapted it.

Thanks for any help,
Adrian

**Comment by dridzal** [comment]
Hi @Adrian-Diaz :  Can you be a little more specific about the piece of code you are referencing?

There may have been an attempt to implement MMA some time ago, for use in topology optimization, but we found considerably better alternatives.  For instance, we performed an in-house study that included, among others, the standard problem of compliance minimization with a volume constraint, and found that projection-based trust-region methods with a _zero Hessian_ in the model (i.e., linear trust-region model) worked as well as optimality criteria (OC) or MMA.  With full Hessian information, i.e., indefinite quadratic models, projection-based trust-region methods outperformed OC and MMA by a wide margin.  For OC and MMA, we used standalone implementations, not through ROL.  Considering these findings, and considering the generality and the strong theoretical foundations of trust-region methods, we decided not to pursue OC or MMA as part of ROL.

I'm CCing @dpkouri and @gregvw , as they may know more about the code you've found.

**Comment by Adrian-Diaz** [comment]
Hey @dridzal 

It's the ROL_ObjectiveMMA.hpp header in src/function. I'm mostly just looking for an expedient way to generate the comparison for academic purposes; especially since it'd be in the case with a different TO approach, multi-objective function, and multiple competing constraints. Is there a published reference for the comparison to MMA you mentioned?

Thanks for the help,

Adrian

**Comment by dridzal** [comment]
I'll let @gregvw and @dpkouri comment on the header.  Unfortunately, the report is not published.  The paper by @dpkouri

https://link.springer.com/article/10.1007/s11590-021-01794-1

has some results on projection-based trust-region methods, from which one could extrapolate a comparison to MMA, based on published MMA results.  If I recall our unpublished report correctly, the Hessian-free (zero-Hessian) version of LMTR performed roughly the same as OC.  So, to get a sense for how well MMA could perform, you could set the Hessian to zero and use the Lin-More method with the volume projection, for the basic TO problem.

**Comment by gregvw** [comment]
Hello @Adrian-Diaz. I wrote that class when we were looking at implementing different approaches to topology optimization. It's been quite a while, but I believe that it implements the MMA objective as it is described in equations 2 through 7 in [The method of moving asymptotes—a new method for structural optimization](https://onlinelibrary.wiley.com/doi/pdf/10.1002/nme.1620240207) by Krister Svanberg, International journal for numerical methods in engineering 24.2 (1987): 359-373. It was mostly written as a proof of concept and I don't recall ever having used it for anything as we did not see any way to analogously implement the MMA form of (in)equality constraints that would work with the abstract ROL::Vector interface.  That being said, I do think it has at least been tested to verify that the gradient and hessVec methods pass finite difference checks. Hope this helps.

Greg
